src/python/espressomd/lb.pyx

```diff
@@ -24,19 +24,160 @@ import functools
 import numpy as np
 cimport numpy as np
 from libc cimport stdint
-from .actors cimport Actor
-from . import utils
 from .utils cimport Vector3i, Vector3d, Vector6d, Vector19d, make_Vector3d
 from .integrate cimport get_time_step
+from . import utils
+from . import highlander
+
+
+cdef class FluidActor:
+
+    """
+    Abstract base class for interactions affecting particles in the system,
+    such as LB fluids. Derived classes must implement the interface to the
+    relevant core objects and global variables.
+    """
+
+    # Keys in active_list have to match the method name.
+    active_list = dict(HydrodynamicInteraction=False)
+
+    # __getstate__ and __setstate__ define the pickle interaction
+    def __getstate__(self):
+        odict = self._params.copy()
+        return odict
+
+    def __setstate__(self, params):
+        self._params = params
+        self._set_params_in_es_core()
+
+    def __init__(self, *args, **kwargs):
+        self._isactive = False
+        utils.check_valid_keys(self.valid_keys(), kwargs.keys())
+        utils.check_required_keys(self.required_keys(), kwargs.keys())
+        self._params = self.default_params()
+        self._params.update(kwargs)
+
+    def _activate(self):
+        inter = self._get_interaction_type()
+        if inter in FluidActor.active_list:
+            if FluidActor.active_list[inter]:
+                raise highlander.ThereCanOnlyBeOne(self.__class__.__bases__[0])
+            FluidActor.active_list[inter] = True
+
+        self.validate_params()
+        self._activate_method()
+        utils.handle_errors("Activation of an actor")
+        self._isactive = True
+
+    def _deactivate(self):
+        self._deactivate_method()
+        utils.handle_errors("Deactivation of an actor")
+        self._isactive = False
+        inter = self._get_interaction_type()
+        if inter in FluidActor.active_list:
+            if not FluidActor.active_list[inter]:
+                raise Exception(
+                    "Class not registered in Actor.active_list: " + self.__class__.__bases__[0].__name__)
```

Comment:
What is the reason we use `+` and `%`-operators for strings instead of f-strings in this whole file?

Reply:
I had moved the original actor code without modifying it. It's now using f-strings everywhere in the latest commit.